File: src/main/java/com/enderzombi102/cmt/command/Executor.java
package com.enderzombi102.cmt.command;

import net.minecraft.entity.Entity;
import net.minecraft.server.MinecraftServer;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.server.function.CommandFunction;
import net.minecraft.server.world.ServerWorld;
import net.minecraft.util.Identifier;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

public class Executor {
	private final @NotNull MinecraftServer server;
	private final @Nullable String command;
	private final @Nullable CommandFunction function;

	public Executor( @NotNull MinecraftServer server, @NotNull String executes ) {
		this.server = server;
		Optional<CommandFunction> func = this.server.getCommandFunctionManager().getFunction( new Identifier(executes) );
		if ( func.isPresent() ) {
			this.function = func.get();
			this.command = null;
		} else {
			this.function = null;
			this.command = executes;
		}
	}

	public void Execute(Entity entity) {
		if ( this.function != null ) {
			this.server.getCommandFunctionManager().execute(
					this.function,
					this.getSource(entity)
			);
		} else {
			this.server.getCommandManager().execute(
				this.getSource(entity),
				this.command
			);
		}
	}

	private ServerCommandSource getSource(Entity entity) {
		return this.server.getCommandSource()
				.withEntity(entity)
				.withWorld( (ServerWorld) entity.getCommandSenderWorld() )
				.withPosition( entity.getPos() )
				.withRotation( entity.getRotationClient() );
	}

}

File: src/main/kotlin/com/enderzombi102/cmt/command/Executor.kt
package com.enderzombi102.cmt.command

import net.minecraft.entity.Entity
import net.minecraft.server.MinecraftServer
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.server.function.CommandFunction
import net.minecraft.server.world.ServerWorld
import net.minecraft.util.Identifier

class Executor( private val server: MinecraftServer, executes: String ) {
	private var command: String? = null
	private var function: CommandFunction? = null

	init {
		val func = server.commandFunctionManager.getFunction( Identifier( executes ) )
		if (func.isPresent) {
			function = func.get()
			command = null
		} else {
			function = null
			command = executes
		}
	}

	fun execute( entity: Entity ) {
		if ( function != null )
			server.commandFunctionManager.execute( function, getSource( entity ) )
		else
			server.commandManager.executePrefixedCommand( getSource( entity ), command )
	}

	private fun getSource(entity: Entity): ServerCommandSource {
		return server.commandSource
			.withEntity(entity)
			.withWorld(entity.commandSenderWorld as ServerWorld)
			.withPosition(entity.pos)
			.withRotation(entity.rotationClient)
	}
}
